clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1957,8 +1960,29 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
   }
 
   bool handleStructType(ParmVarDecl *PD, QualType ParamTy) final {
-    Diag.Report(PD->getLocation(), diag::err_bad_kernel_param_type) << ParamTy;
-    IsInvalid = true;
+    CXXRecordDecl *RD = ParamTy->getAsCXXRecordDecl();
+    // For free functions all struct/class kernel arguments are forward declared
+    // in integration header, that adds additional restrictions for kernel
+    // arguments.
+    // Lambdas are not forward declarable. So, diagnose them properly.
+    if (RD->isLambda()) {
+      Diag.Report(PD->getLocation(), diag::err_bad_kernel_param_type)
+          << ParamTy;
+      IsInvalid = true;
+      return isValid();
+    }
+
+    // Check that the type is defined at namespace scope.
+    const DeclContext *DeclCtx = RD->getDeclContext();
+    while (!DeclCtx->isTranslationUnit() &&
+           (isa<NamespaceDecl>(DeclCtx) || isa<LinkageSpecDecl>(DeclCtx)))
+      DeclCtx = DeclCtx->getParent();
+
+    if (!DeclCtx->isTranslationUnit()) {
+      Diag.Report(PD->getLocation(), diag::err_bad_kernel_param_type)
+          << ParamTy;
+      IsInvalid = true;
+    }
```

Comment:
Perhaps it would make sense to factor this out to an `isForwardDeclarable()` function? I looked for an existing one, but didn't find one. There are several cases where a forward declarable declaration is required and I'm already skeptical that we're diagnosing violations correctly. See `DiagnoseKernelNameType()` for additional code that could be factored out and merged.

Reply:
Done.